include/rapidcheck/state/gen/ParallelCommands.hpp

```diff
@@ -0,0 +1,341 @@
+    return shrinkable::shrinkRecur(generateInitialParallel(random, size),
+                                   &shrinkSequence);
+  }
+
+  ParallelCommandSequence generateInitialParallel(const Random &random,
+                                                  int size) const {
+    auto r = random;
+    int prefixSz, leftSz, rightSz;
+    std::tie(prefixSz, leftSz, rightSz) = parallelCommandDistribution(size);
+
+    std::size_t prefixCount = (r.split().next() % (prefixSz + 1));
+    std::size_t leftCount = (r.split().next() % (leftSz + 1));
+    std::size_t rightCount = (r.split().next() % (rightSz + 1));
+
+    return ParallelCommandSequence(
+        generateInitial(random, prefixSz, prefixCount),
+        generateInitial(random, leftSz, leftCount),
+        generateInitial(random, rightSz, rightCount));
+  }
+
+  Shrinkable<CommandSequence> generateSequence(const Random &random,
+                                               int size) const {
+    Random r(random);
+    std::size_t count = (r.split().next() % (size + 1)) + 1;
+    return shrinkable::shrinkRecur(generateInitial(random, size, count),
+                                   &shrinkSequence);
+  }
+
+  CommandSequence
+  generateInitial(const Random &random, int size, std::size_t count) const {
+    CommandSequence sequence(m_genFunc, size);
+    sequence.entries.reserve(count);
+
+    auto r = random;
+    while (sequence.entries.size() < count) {
+      sequence.entries.push_back(nextEntry(r.split(), size));
+    }
+
+    return sequence;
+  }
+
+  CommandEntry nextEntry(const Random &random, int size) const {
+    using namespace ::rc::detail;
+    auto r = random;
+    const auto gen = m_genFunc();
+    // TODO configurability?
+    for (int tries = 0; tries < 100; tries++) {
+      try {
+        auto random = r.split();
+        auto shrinkable = gen(random, size);
+
+        return CommandEntry(std::move(random), std::move(shrinkable));
+      } catch (const CaseResult &result) {
+        if (result.type != CaseResult::Type::Discard) {
+          throw;
+        }
+        // What to do?
+      } catch (const GenerationFailure &failure) {
+        // What to do?
+      }
+    }
+
+    // TODO better error message
+    throw GenerationFailure("Failed to generate command after 100 tries.");
+  }
+
+  static Seq<ParallelCommandSequence>
+  shrinkSequence(const ParallelCommandSequence &s) {
+    return seq::concat(shrinkPrefix(s),
+                       shrinkLeft(s),
+                       shrinkRight(s),
+                       unparallelizeLeft(s),
+                       unparallelizeRight(s));
+  }
+
+  static Seq<ParallelCommandSequence>
+  unparallelizeLeft(const ParallelCommandSequence &s) {
+    auto valid = seq::filter(seq::just(s),
+                             [](const ParallelCommandSequence &s) {
+                               return !s.left.entries.empty();
+                             });
+
+    return seq::map(std::move(valid),
```

Comment:
New stragety implemented in db31d55560396e43a3115a15a143762cae028fc1.

Not sure that this is the best strategy however since it generates more shrinks than maybe is necessary. The strategy is to start by serializing both left and right, e.g. prefix=ABC left=DEF right=GHI is mapped to prefix=ABCDEFGHI. Then one element at a time is moved back to its origin branch. Different orderings are however not tested (commands from left are always placed before commands from right).

It might be better to take one branch at a time instead.